watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/dye_desalination/dye_desalination.py

```diff
@@ -0,0 +1,329 @@
+import idaes.core.util.scaling as iscale
+from idaes.generic_models.costing import UnitModelCostingBlock
+
+from watertap.core.util.initialization import assert_degrees_of_freedom
+
+from watertap.core.wt_database import Database
+import watertap.core.zero_order_properties as prop_ZO
+from watertap.unit_models.zero_order import (
+    FeedZO,
+    PumpElectricityZO,
+    NanofiltrationZO,
+)
+from watertap.core.zero_order_costing import ZeroOrderCosting
+
+
+def main():
+    m = build()
+    set_operating_conditions(m)
+    assert_degrees_of_freedom(m, 0)
+    assert_units_consistent(m)
+
+    results = solve(m)
+    assert_optimal_termination(results)
+
+    display_results(m)
+
+    add_costing(m)
+    m.fs.costing.initialize()
+
+    results = solve(m)
+    assert_optimal_termination(results)
+    display_costing(m)
+    return m, results
+
+
+def build():
+    # flowsheet set up
+    m = ConcreteModel()
+    m.db = Database()
+
+    m.fs = FlowsheetBlock(default={"dynamic": False})
+    m.fs.prop = prop_ZO.WaterParameterBlock(default={"solute_list": ["dye", "tds"]})
+
+    # unit model
+    m.fs.feed = FeedZO(default={"property_package": m.fs.prop})
+
+    m.fs.P1 = PumpElectricityZO(
+        default={
+            "property_package": m.fs.prop,
+            "database": m.db,
+            "process_subtype": "default",
+        }
+    )
+
+    m.fs.nanofiltration = NanofiltrationZO(
+        default={
+            "property_package": m.fs.prop,
+            "database": m.db,
+            "process_subtype": "rHGO_dye_rejection",
+        }
+    )
+
+    m.fs.permeate1 = Product(default={"property_package": m.fs.prop})
+    m.fs.retentate1 = Product(default={"property_package": m.fs.prop})
+
+    # connections
+    m.fs.s01 = Arc(source=m.fs.feed.outlet, destination=m.fs.P1.inlet)
+    m.fs.s02 = Arc(source=m.fs.P1.outlet, destination=m.fs.nanofiltration.inlet)
+    m.fs.s03 = Arc(source=m.fs.nanofiltration.treated, destination=m.fs.permeate1.inlet)
+    m.fs.s04 = Arc(
+        source=m.fs.nanofiltration.byproduct, destination=m.fs.retentate1.inlet
+    )
+
+    TransformationFactory("network.expand_arcs").apply_to(m)
+
+    # scaling
+    iscale.calculate_scaling_factors(m)
+
+    return m
+
+
+def set_operating_conditions(m):
+    # feed
+    flow_vol = 120 / 3600 * pyunits.m**3 / pyunits.s
+    conc_mass_dye = 2.5 * pyunits.kg / pyunits.m**3
+    conc_mass_tds = 50.0 * pyunits.kg / pyunits.m**3
+
+    m.fs.feed.flow_vol[0].fix(flow_vol)
+    m.fs.feed.conc_mass_comp[0, "dye"].fix(conc_mass_dye)
+    m.fs.feed.conc_mass_comp[0, "tds"].fix(conc_mass_tds)
+    solve(m.fs.feed)
+
+    # nanofiltration
+    m.fs.nanofiltration.load_parameters_from_database(use_default_removal=True)
+
+    # pump
+    m.fs.P1.load_parameters_from_database(use_default_removal=True)
+    m.fs.P1.applied_pressure.fix(m.fs.nanofiltration.applied_pressure.get_values()[0])
+    m.fs.P1.lift_height.unfix()
+    initialize_system(m)
```

Comment:
I noticed that our other finished flowsheets call `initialize_system(m)` from `main()`, instead of at the end of `set_operating_conditions(m)` as you have done here. Does it make a difference? I think not, but I am not 100% sure.

Reply:
As you point out, I am not sure that it makes a difference functionally, but I changed it to match the other flowsheets.